azure-pipelines.yaml

```diff
@@ -0,0 +1,201 @@
+    azureContainerRegistry: $(azureContainerRegistryFull)
+    dockerComposeFile: docker-compose.yaml
+    projectName: $(Build.Repository.Name)
+    qualifyImageNames: true
+    dockerComposeCommand: -f docker-compose.test.yaml build --no-cache $(imageName)
+  displayName: build tests image
+  # run tests
+- task: DockerCompose@0
+  inputs:
+    containerregistrytype: Azure Container Registry
+    azureSubscriptionEndpoint: $(azureSubscriptionEndpoint)
+    azureContainerRegistry: $(azureContainerRegistryFull)
+    dockerComposeFile: docker-compose.yaml
+    projectName: $(Build.Repository.Name)
+    qualifyImageNames: true
+    dockerComposeCommand: -f docker-compose.test.yaml up --force-recreate  $(imageName)
+  displayName: run tests
+  # build base image
+- task: DockerCompose@0
+  inputs:
+    action: Build services
+    containerregistrytype: Azure Container Registry
+    azureSubscriptionEndpoint: $(azureSubscriptionEndpoint)
+    azureContainerRegistry: $(azureContainerRegistryFull)
+    dockerComposeFile: docker-compose.yaml
+    projectName: $(Build.Repository.Name)
+    qualifyImageNames: true
+    additionalImageTags: $(containerTag)
+  displayName: build base image
+  # push container to registry
+- task: DockerCompose@0
+  inputs:
+    action: Push services
+    azureSubscriptionEndpoint: $(azureSubscriptionEndpoint)
+    azureContainerRegistry: $(azureContainerRegistryFull)
+    dockerComposeFile: docker-compose.yaml
+    qualifyImageNames: true
+    additionalImageTags: $(containerTag)
+  displayName: Push container to registry
+  # publish test results
+- task: PublishTestResults@2
+  displayName: Publish test results
+  inputs:
+    testResultsFormat: JUnit
+    testResultsFiles: test-output/junit.xml
+    failTaskOnFailedTests: true
+    testRunTitle: Unit Test
+  # deploy via helm
+- task: HelmDeploy@0
+  inputs:
+    azureSubscriptionEndpoint: $(azureSubscriptionEndpoint)
+    azureResourceGroup: $(azureResourceGroup)
+    kubernetesCluster: $(kubernetesCluster)
+    namespace: $(imageName)-$(safeContainerTag)
+    command: upgrade
+    chartType: filepath
+    chartPath: helm
+    releaseName: $(imageName)-$(containerTag)
+    overrideValues: 'container.messageQueueHost=$(messageQueueHost),container.messageQueuePort=$(messageQueuePort),container.messageQueueTransport=$(messageQueueTransport),container.scheduleQueueAddress=$(scheduleQueueAddress),container.scheduleQueueUser=$(scheduleQueueUser),container.scheduleQueuePassword=$(scheduleQueuePassword),container.paymentQueueAddress=$(paymentQueueAddress),container.paymentQueueUser=$(paymentQueueUser),container.paymentQueuePassword=$(paymentQueuePassword),postgresExternalName=$(postgresExternalName),postgresPassword=$(postgresPassword)'
+    arguments: --set image=$(azureContainerRegistryFull)/$(imageName):$(containerTag) --set container.redeployOnChange=$(Build.BuildId) --set container.imagePullPolicy=Always --atomic

```

Comment:
I haven't tested this but yaml allows for [multi-line formatting](https://yaml-multiline.info/) where newlines are replaced by spaces, so I think the long `arguments` line could be made easier to read:
```
arguments: >
  --set image=$(azureContainerRegistryFull)/$(imageName):$(containerTag)
  --set container.redeployOnChange=$(Build.BuildId)
  --set container.imagePullPolicy=Always
  --atomic
```

Reply:
Interesting, I didn't know that.  I'll leave as is for now but will experiment for future changes.